vortex-filtering/include/vortex_filtering/filters/ekf.hpp

```diff
@@ -0,0 +1,135 @@
+/**
+ * @file ekf.hpp
+ * @author Eirik Kolås
+ * @brief Extended Kalman Filter.
+ * @version 0.1
+ * @date 2023-10-26
+ *
+ * @copyright Copyright (c) 2023
+ */
+
+#pragma once
+#include <tuple>
+#include <vortex_filtering/models/dynamic_model_interfaces.hpp>
+#include <vortex_filtering/models/sensor_model_interfaces.hpp>
+#include <vortex_filtering/probability/multi_var_gauss.hpp>
+
+namespace vortex::filter {
+
+/**
+ * @brief Extended Kalman Filter.
+ * @tparam DynModT Dynamic model type derived from `vortex::models::interface::DynamicModelLTV`
+ * @tparam SensModT Sensor model type derived from `vortex::models::interface::SensorModelLTV`
+ */
+template <models::concepts::DynamicModelLTV DynModT, models::concepts::SensorModelLTV SensModT> class EKF {
+public:
+  static constexpr int N_DIM_x = DynModT::DynModI::N_DIM_x;
+  static constexpr int N_DIM_z = SensModT::SensModI::N_DIM_z;
+  static constexpr int N_DIM_u = DynModT::DynModI::N_DIM_u;
+  static constexpr int N_DIM_v = DynModT::DynModI::N_DIM_v;
+  static constexpr int N_DIM_w = SensModT::SensModI::N_DIM_w;
+
+  using Vec_x = Eigen::Vector<double, N_DIM_x>;
+  using Vec_z = Eigen::Vector<double, N_DIM_z>;
+  using Vec_u = Eigen::Vector<double, N_DIM_u>;
+  using Vec_v = Eigen::Vector<double, N_DIM_v>;
+  using Vec_w = Eigen::Vector<double, N_DIM_w>;
+
+  using Mat_xx = Eigen::Matrix<double, N_DIM_x, N_DIM_x>;
+  using Mat_xz = Eigen::Matrix<double, N_DIM_x, N_DIM_z>;
+  using Mat_xv = Eigen::Matrix<double, N_DIM_x, N_DIM_v>;
+  using Mat_xw = Eigen::Matrix<double, N_DIM_x, N_DIM_w>;
+
+  using Mat_zx = Eigen::Matrix<double, N_DIM_z, N_DIM_x>;
+  using Mat_zz = Eigen::Matrix<double, N_DIM_z, N_DIM_z>;
+  using Mat_zw = Eigen::Matrix<double, N_DIM_z, N_DIM_w>;
+
+  using Mat_vx = Eigen::Matrix<double, N_DIM_v, N_DIM_x>;
+  using Mat_vv = Eigen::Matrix<double, N_DIM_v, N_DIM_v>;
+  using Mat_vw = Eigen::Matrix<double, N_DIM_v, N_DIM_w>;
+
+  using Mat_wx = Eigen::Matrix<double, N_DIM_w, N_DIM_x>;
+  using Mat_wv = Eigen::Matrix<double, N_DIM_w, N_DIM_v>;
+  using Mat_ww = Eigen::Matrix<double, N_DIM_w, N_DIM_w>;
+
+  using Gauss_x = prob::MultiVarGauss<N_DIM_x>;
+  using Gauss_z = prob::MultiVarGauss<N_DIM_z>;
+  using Gauss_v = prob::MultiVarGauss<N_DIM_v>;
+  using Gauss_w = prob::MultiVarGauss<N_DIM_w>;
+
+  using DynModI  = models::interface::DynamicModelLTV<N_DIM_x, N_DIM_u, N_DIM_v>;
+  using SensModI = models::interface::SensorModelLTV<N_DIM_x, N_DIM_z, N_DIM_w>;
+
+  using DynModIPtr  = typename DynModI::SharedPtr;
+  using SensModIPtr = typename SensModI::SharedPtr;
+
+  /** Construct a new EKF object.
+   * @tparam DynamicModelT Dynamic model type derived from `vortex::models::interface::DynamicModelLTV`
+   * @tparam SensorModelT Sensor model type derived from `vortex::models::interface::SensorModelLTV`
+   */
+  EKF() = default;
+
```

Comment:
Just food for thought (as both methods can work well), what was the reason for templating across dynamic- and sensor models instead of doing dependency injection with the constructor?

With injection, I suspect you could simplify the method signatures and further simplify with a factory for the EKF class

Reply:
The reason for templating across the models is to allow the different sizes of states, measurements, inputs, noises etc. to be determined at compile-time. Having all the mothods static makes it so you never have to create an instance of the class, which I found convenient. 

I also didn't want to store the models inside the EKF, as updating Q and R would have to be bucket-brigadiered (as Sverre would put it) through both the EKF and the model.

Reply:
Ahh I see. The EKF and UKF were originally templated on the state sizes (of x, z, u, v, w) with an alias to make it also templatable with the models (by just extracting the sizes from the models). 

The reason it is the way it is now is to abstract away having to think about the sizes of five vectors. The intended way to use it is to create your own alias of the ekf with the models, and then you don't have to think about the template parameters anyway. I get that that is probably a non-standard way to create an api. What do you think?

Reply:
Honestly I think this specific design is really cool from the user-perspective. A bit hard to read from the implementation side, but that's fine (just took me a few read-throughs to see the patterns). 

IMO, keep things the way they are now, and only change things based on use-case experience in the tracker and such.